File: chapter-01/my_set.rb
class MySet
    attr_reader :items

    def initialize()
        @items = []
    end

    def push(item)
        items.push(item) unless items.include?(item)
    end
end

set = MySet.new
set.push("apple")
puts "Initialized set: #{set.items}"

new_set = MySet.new
new_set.push("pie")
puts "Initialized new_set: #{new_set.items}"

set.push("apple")
puts "Attempted to add \"apple\" to set with result: #{set.items}"

new_set.push("apple")
puts "Attempted to add \"apple\" to new_set with result: #{new_set.items}"

File: chapter-01/my_set.ex
defmodule MySet do
  defstruct items: []

  def push(set = %{items: items}, item) do
    if Enum.member?(items, item) do
      set
    else
      %{set | items: items ++ [item]}
    end
  end
end

# Commands to use in IEx session
# c("my_set.ex")
# set = %MySet{}
# set = MySet.push(set, "apple")
# IO.inspect set

# new_set = %MySet{}
# new_set = MySet.push(new_set, "pie")
# IO.inspect new_set

# set = MySet.push(set, "apple")
# IO.inspect set

# new_set = MySet.push(new_set, "apple")
# IO.inspect new_set
